pip/pip-297.md

```diff
@@ -0,0 +1,168 @@
+the notifyError method to the PushSource class in PIP-281: https://github.com/apache/pulsar/pull/20807. However, this
+does not solve the same problem that all source connectors face.
+
+The problem is same for the Pulsar Function. The user can also process messages asynchronously in a Pulsar Function. If
+any fatal exceptions happen in the user's thread, the function framework can't handle the exception
+correctly.
+
+We need a way for the connector and function developers to throw unrecoverable exceptions outside the function instance
+thread. The function framework should catch these exceptions and terminate the function accordingly.
+
+# Goals
+
+## In Scope
+
+- This proposal will apply both to the Pulsar Function and the Pulsar Connector.
+- Support throwing unrecoverable exceptions outside the function instance thread.
+- The function framework should terminate the connector or function when there are any unrecoverable exceptions thrown
+  asynchronously.
+
+## Out of Scope
+
+- The fixes of the exception-raising issue mentioned in the Motivation part for all the connectors are not included in
+  this PIP. This PIP only provides the feature for the connector developer to raise the exception outside the function
+  instance thread. The fixes should be in serval different PRs.
+
+# High Level Design
+
+Introduce a new method `raiseException` to the context. All the connector implementation code and the function code can
+use this context and call the `raiseException` method to raise an unrecoverable exception to the function framework.
+
+After the connector or function raises the unrecoverable exception, the function instance thread will be interrupted.
+The function framework then could catch the exception, log it, and then terminate the function instance.
+
+# Detailed Design
+
+## Design & Implementation Details
+
+This PIP proposes to add a new method`raiseException`to the context `BaseContext`. This method allows the
+connector or the function code to report an unrecoverable exception to the function framework. The `SinkContext`
+and `SourceContext`
+are all inherited from `BaseContext`. Therefore, all the sink connectors and source connectors can invoke this new
+method. The pulsar function context class `Context` is also inherited from `BaseContext`. Therefore, the function code
+can also invoke this new method.
+
+In the `raiseException`, the function instance thread will be interrupted. The function instance thread can then catch
+the interrupt exception and get the unrecoverable exception. The function framework then logs this exception,
+reports to the metrics, and finally terminates the function instance.
+
+## Public-facing Changes
+
+### Public API
+
+Introduce `raiseException` method to the `BaseContext`:
+
+```java
+public interface BaseContext {
+    /**
+     * Raise an unrecoverable exception to the function framework.
+     *
+     * @param e the exception to be raised
+     */
+    void raiseException(Exception e);
```

Comment:
How does the function framework know if the exception is a retriable exception or an un-retriable exception? From the description. It looks like the connector or function should only broadcast the un-retriable exception by this method. If yes, maybe we'd better change the method name to raiseUnRetriableException().

And if there is a way to terminate the Function Instance. IMO, it should be better than raising exceptions.

Reply:
Good suggestion. This method should only handle `un-retriable exception` or, to be clearer, `fatal exception`. 

I have renamed it to `raiseFatalException`.